src/ext/java/org/opentripplanner/ext/siri/updater/AsyncEstimatedTimetableProcessor.java

```diff
@@ -0,0 +1,45 @@
+package org.opentripplanner.ext.siri.updater;
+
+import java.util.concurrent.Future;
+import java.util.function.Consumer;
+import org.opentripplanner.updater.spi.UpdateResult;
+import org.opentripplanner.updater.spi.WriteToGraphCallback;
+import org.opentripplanner.updater.trip.UpdateIncrementality;
+import uk.org.siri.siri20.ServiceDelivery;
+
+/**
+ * Apply asynchronously estimated timetable updates in the graph-writer thread and forward the
+ * result to an update result consumer.
+ */
+public class AsyncEstimatedTimetableProcessor {
+
+  private final EstimatedTimetableHandler estimatedTimetableHandler;
+  private final WriteToGraphCallback saveResultOnGraph;
+  private final Consumer<UpdateResult> updateResultConsumer;
+
+  public AsyncEstimatedTimetableProcessor(
+    EstimatedTimetableHandler estimatedTimetableHandler,
+    WriteToGraphCallback saveResultOnGraph,
+    Consumer<UpdateResult> updateResultConsumer
+  ) {
+    this.estimatedTimetableHandler = estimatedTimetableHandler;
+    this.saveResultOnGraph = saveResultOnGraph;
+    this.updateResultConsumer = updateResultConsumer;
+  }
+
+  /**
+   * Apply the estimated timetables to the transit model.
+   * This method is non-blocking and applies the changes asynchronously.
+   * @return a future indicating when the changes are applied.
+   */
+  public Future<?> processSiriData(ServiceDelivery serviceDelivery) {
```

Comment:
What are you trying to achieve by using `java.util.concurrent.Future`? I always though that it's pretty useless since it `get` is a blocking call.

`CompletableFuture` has a much better model and truly enables async programming. Is this needed here?

However, with [Structured Concurrency](https://openjdk.org/jeps/453) I don't really know how much this is going to matter anymore. Do you?

Reply:
We need that blocking call only when initializing the updater: the updater is primed when all the messages in the message cache/message history are applied. (that is why in `GooglePubsubEstimatedTimetableSource` we set `primed` to `true` only when future.get() returns)
After initialization, the return value of processSiriData() is simply ignored since we want the updates to be applied asynchronously, in a non-blocking way.
We do not need a full-fledged `CompletableFuture` since the source does not need to know the actual result of the update. It just needs to know that the update is complete.